Index: apc_main.c
===================================================================
--- apc_main.c	(Revision 311029)
+++ apc_main.c	(Revision 324625)
@@ -391,7 +391,7 @@
 /* }}} */
 
 /* {{{ apc_compile_cache_entry  */
-zend_bool apc_compile_cache_entry(apc_cache_key_t key, zend_file_handle* h, int type, time_t t, zend_op_array** op_array, apc_cache_entry_t** cache_entry TSRMLS_DC) {
+zend_bool apc_compile_cache_entry(apc_cache_key_t *key, zend_file_handle* h, int type, time_t t, zend_op_array** op_array, apc_cache_entry_t** cache_entry TSRMLS_DC) {
     int num_functions, num_classes;
     apc_function_t* alloc_functions;
     zend_op_array* alloc_op_array;
@@ -437,7 +437,7 @@
             while((n = php_stream_read(stream, (char*)buf, sizeof(buf))) > 0) {
                 PHP_MD5Update(&context, buf, n);
             }
-            PHP_MD5Final(key.md5, &context);
+            PHP_MD5Final(key->md5, &context);
             php_stream_close(stream);
             if(n<0) {
                 apc_warning("Error while reading '%s' for md5 generation." TSRMLS_CC, filename);
@@ -459,7 +459,7 @@
     }
 
     path = h->opened_path;
-    if(!path && key.type == APC_CACHE_KEY_FPFILE) path = (char*)key.data.fpfile.fullpath;
+    if(!path && key->type == APC_CACHE_KEY_FPFILE) path = (char*)key->data.fpfile.fullpath;
     if(!path) path=h->filename;
 
     apc_debug("2. h->opened_path=[%s]  h->filename=[%s]\n" TSRMLS_CC, h->opened_path?h->opened_path:"null",h->filename);
@@ -607,7 +607,7 @@
 #endif
 
     zend_try {
-        if (apc_compile_cache_entry(key, h, type, t, &op_array, &cache_entry TSRMLS_CC) == SUCCESS) {
+        if (apc_compile_cache_entry(&key, h, type, t, &op_array, &cache_entry TSRMLS_CC) == SUCCESS) {
             ctxt.pool = cache_entry->pool;
             ctxt.copy = APC_COPY_IN_OPCODE;
             if (apc_cache_insert(apc_cache, key, cache_entry, &ctxt, t TSRMLS_CC) != 1) {
@@ -842,8 +842,10 @@
 #endif
 
 #ifdef ZEND_ENGINE_2_4
+#ifndef ZTS
     apc_interned_strings_init(TSRMLS_C);
 #endif
+#endif
 
     APCG(initialized) = 1;
     return 0;
@@ -887,14 +889,16 @@
         apc_cache_release(apc_cache, cache_entry TSRMLS_CC);
     }
 
+#ifdef ZEND_ENGINE_2_4
+#ifndef ZTS
+    apc_interned_strings_shutdown(TSRMLS_C);
+#endif
+#endif
+
     apc_cache_destroy(apc_cache TSRMLS_CC);
     apc_cache_destroy(apc_user_cache TSRMLS_CC);
     apc_sma_cleanup(TSRMLS_C);
 
-#ifdef ZEND_ENGINE_2_4
-    apc_interned_strings_shutdown(TSRMLS_C);
-#endif
-
     APCG(initialized) = 0;
     return 0;
 }
Index: apc_compile.c
===================================================================
--- apc_compile.c	(Revision 311029)
+++ apc_compile.c	(Revision 324625)
@@ -33,14 +33,11 @@
 #include "apc_compile.h"
 #include "apc_globals.h"
 #include "apc_zend.h"
+#include "apc_php.h"
 #include "apc_string.h"
 #include "ext/standard/php_var.h"
 #include "ext/standard/php_smart_str.h"
 
-#ifndef IS_CONSTANT_TYPE_MASK
-#define IS_CONSTANT_TYPE_MASK (~IS_CONSTANT_INDEX)
-#endif
-
 typedef void* (*ht_copy_fun_t)(void*, void*, apc_context_t* TSRMLS_DC);
 //typedef void  (*ht_free_fun_t)(void*, apc_context_t*);
 typedef int (*ht_check_copy_fun_t)(Bucket*, va_list);
@@ -60,7 +57,9 @@
  */
 static zval** my_copy_zval_ptr(zval**, const zval**, apc_context_t* TSRMLS_DC);
 static zval* my_copy_zval(zval*, const zval*, apc_context_t* TSRMLS_DC);
+#ifndef ZEND_ENGINE_2_4
 static znode* my_copy_znode(znode*, znode*, apc_context_t* TSRMLS_DC);
+#endif
 static zend_op* my_copy_zend_op(zend_op*, zend_op*, apc_context_t* TSRMLS_DC);
 static zend_function* my_copy_function(zend_function*, zend_function*, apc_context_t* TSRMLS_DC);
 static zend_function_entry* my_copy_function_entry(zend_function_entry*, const zend_function_entry*, apc_context_t* TSRMLS_DC);
@@ -93,9 +92,11 @@
  * defined/overridden in the 'current' class and not inherited.
  */
 static int my_check_copy_function(Bucket* src, va_list args);
+static int my_check_copy_property_info(Bucket* src, va_list args);
+#ifndef ZEND_ENGINE_2_4
 static int my_check_copy_default_property(Bucket* p, va_list args);
-static int my_check_copy_property_info(Bucket* src, va_list args);
 static int my_check_copy_static_member(Bucket* src, va_list args);
+#endif
 static int my_check_copy_constant(Bucket* src, va_list args);
 
 /* }}} */
@@ -278,13 +279,15 @@
 static char *apc_string_pmemcpy(char *str, size_t len, apc_pool* pool TSRMLS_DC)
 {	
 #ifdef ZEND_ENGINE_2_4
+#ifndef ZTS
     if (pool->type != APC_UNPOOL) {
-        char * ret = apc_new_interned_string(str, len TSRMLS_CC);
+        char * ret = (char*)apc_new_interned_string((const char*)str, len TSRMLS_CC);
         if (ret) {
             return ret;
         }
     }
 #endif
+#endif
     return apc_pmemcpy(str, len, pool TSRMLS_CC);
 }
 
@@ -886,20 +889,21 @@
             CHECK((newp = (Bucket*) apc_pmemcpy(curr, sizeof(Bucket), pool TSRMLS_CC)));
         } else if (IS_INTERNED(curr->arKey)) {
             CHECK((newp = (Bucket*) apc_pmemcpy(curr, sizeof(Bucket), pool TSRMLS_CC)));
+#ifndef ZTS
         } else if (pool->type != APC_UNPOOL) {
             char *arKey;
 
+            CHECK((newp = (Bucket*) apc_pmemcpy(curr, sizeof(Bucket), pool TSRMLS_CC)));
             arKey = apc_new_interned_string(curr->arKey, curr->nKeyLength TSRMLS_CC);
             if (!arKey) {
-                CHECK((newp = (Bucket*) apc_pmemcpy(curr, (sizeof(Bucket) + curr->nKeyLength), pool TSRMLS_CC)));
-                newp->arKey = ((char*)newp) + sizeof(Bucket);
+                CHECK((newp->arKey = (char*) apc_pmemcpy(curr->arKey, curr->nKeyLength, pool TSRMLS_CC)));
             } else {
-                CHECK((newp = (Bucket*) apc_pmemcpy(curr, sizeof(Bucket), pool TSRMLS_CC)));
                 newp->arKey = arKey;
             }
+#endif
         } else {
-            CHECK((newp = (Bucket*) apc_pmemcpy(curr, (sizeof(Bucket) + curr->nKeyLength), pool TSRMLS_CC)));
-            newp->arKey = ((char*)newp) + sizeof(Bucket);
+            CHECK((newp = (Bucket*) apc_pmemcpy(curr, sizeof(Bucket), pool TSRMLS_CC)));
+            CHECK((newp->arKey = (char*) apc_pmemcpy(curr->arKey, curr->nKeyLength, pool TSRMLS_CC)));
         }        
 #else
         CHECK((newp = (Bucket*) apc_pmemcpy(curr,
@@ -947,6 +951,8 @@
 
     dst->pListTail = newp;
 
+    zend_hash_internal_pointer_reset(dst);
+
     return dst;
 }
 /* }}} */
@@ -1016,6 +1022,9 @@
             case ZEND_JMP_SET:
 #endif
 #ifdef ZEND_ENGINE_2_4
+            case ZEND_JMP_SET_VAR:
+#endif
+#ifdef ZEND_ENGINE_2_4
                 zo->op2.jmp_addr = dst->opcodes + (zo->op2.jmp_addr - src->opcodes);
 #else
                 zo->op2.u.jmp_addr = dst->opcodes + (zo->op2.u.jmp_addr - src->opcodes);
@@ -1125,6 +1134,9 @@
 #ifdef ZEND_ENGINE_2_3
             case ZEND_JMP_SET:
 #endif
+#ifdef ZEND_ENGINE_2_4
+            case ZEND_JMP_SET_VAR:
+#endif
                 if(flags != NULL) {
                     flags->has_jumps = 1;
                 }
@@ -1582,6 +1594,9 @@
                 case ZEND_JMP_SET:
 #endif
 #ifdef ZEND_ENGINE_2_4
+                case ZEND_JMP_SET_VAR:
+#endif
+#ifdef ZEND_ENGINE_2_4
                     dzo->op2.jmp_addr = dst->opcodes +
                                             (zo->op2.jmp_addr - src->opcodes);
 #else
@@ -1694,6 +1709,7 @@
     /* Deep-copy the class properties, because they will be modified */
 
 #ifdef ZEND_ENGINE_2_4
+    dst->name = apc_string_pmemcpy((char*)src->name, src->name_length+1, ctxt->pool TSRMLS_CC); 
 	dst->default_properties_count = src->default_properties_count;
     if (src->default_properties_count) {
         dst->default_properties_table = (zval**) apc_php_malloc((sizeof(zval*) * src->default_properties_count) TSRMLS_CC);
Index: apc_iterator.c
===================================================================
--- apc_iterator.c	(Revision 311029)
+++ apc_iterator.c	(Revision 324625)
@@ -118,7 +118,7 @@
     }
     if (APC_ITER_MD5 & iterator->format) {
         if(slot->value->type == APC_CACHE_ENTRY_FILE) {
-            if(slot->key.md5) {
+            if(APCG(file_md5) && slot->key.md5) {
                 make_digest(md5str, slot->key.md5);
                 add_assoc_string(item->value, "md5", md5str, 1);
             }
@@ -217,6 +217,9 @@
     iterator->obj.ce = ce;
     ALLOC_HASHTABLE(iterator->obj.properties);
     zend_hash_init(iterator->obj.properties, 0, NULL, ZVAL_PTR_DTOR, 0);
+#ifdef ZEND_ENGINE_2_4
+    iterator->obj.properties_table = NULL;
+#endif
     iterator->obj.guards = NULL;
     iterator->initialized = 0;
     retval.handle = zend_objects_store_put(iterator, apc_iterator_destroy, apc_iterator_free, NULL TSRMLS_CC);
Index: apc_windows_srwlock_kernel.c
===================================================================
--- apc_windows_srwlock_kernel.c	(Revision 311029)
+++ apc_windows_srwlock_kernel.c	(Revision 324625)
@@ -81,7 +81,15 @@
 
 void apc_windows_cs_destroy(apc_windows_cs_rwlock_t *lock)
 {
-    pRtlDeleteResource(lock);
+    __try
+    {
+        pRtlDeleteResource(lock);
+    }
+        __except(GetExceptionCode() == EXCEPTION_ACCESS_VIOLATION ?
+               EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH)
+    {
+        /* Ignore exception (resource was freed during shutdown of another thread) */
+    }
     FreeLibrary(ntdll);
     return;
 }
Index: apc_string.c
===================================================================
--- apc_string.c	(Revision 311029)
+++ apc_string.c	(Revision 324625)
@@ -25,7 +25,7 @@
 
  */
 
-/* $Id: $ */
+/* $Id$ */
 
 #include "apc.h"
 #include "apc_globals.h"
@@ -34,6 +34,7 @@
 
 #ifdef ZEND_ENGINE_2_4
 
+#ifndef ZTS
 typedef struct _apc_interned_strings_data_t {
     char *interned_strings_start;
     char *interned_strings_end;
@@ -64,9 +65,11 @@
 static void apc_dummy_interned_strings_restore_for_php(TSRMLS_D)
 {
 }
+#endif
 
-char *apc_new_interned_string(char *arKey, int nKeyLength TSRMLS_DC)
+const char *apc_new_interned_string(const char *arKey, int nKeyLength TSRMLS_DC)
 {
+#ifndef ZTS
     ulong h;
     uint nIndex;
     Bucket *p;
@@ -124,8 +127,12 @@
     APCSG(interned_strings).nNumOfElements++;
 
     return p->arKey;
+#else
+    return zend_new_interned_string(arKey, nKeyLength, 0 TSRMLS_CC);
+#endif
 }
 
+#ifndef ZTS
 static void apc_copy_internal_strings(TSRMLS_D)
 {
     Bucket *p, *q;
@@ -146,6 +153,10 @@
             p->arKey = apc_new_interned_string(p->arKey, p->nKeyLength TSRMLS_CC);
         }
 
+		if (ce->name) {
+            ce->name = apc_new_interned_string(ce->name, ce->name_length TSRMLS_CC);
+		}
+
         q = ce->properties_info.pListHead;
         while (q) {
             zend_property_info *info = (zend_property_info*)(q->pData);
@@ -155,7 +166,7 @@
             }
 
             if (info->name) {
-                info->name = apc_new_interned_string(info->name, info->name_length+1 TSRMLS_CC);
+                info->name = apc_new_interned_string(info->name, info->name_length TSRMLS_CC);
             }
 
             q = q->pListNext;
@@ -194,6 +205,7 @@
     int count = APCG(shm_strings_buffer) / (sizeof(Bucket) + sizeof(Bucket*) * 2);
 
     apc_interned_strings_data = (apc_interned_strings_data_t*) apc_sma_malloc(APCG(shm_strings_buffer) TSRMLS_CC);
+    memset((void *)apc_interned_strings_data, 0, APCG(shm_strings_buffer));
 
     CREATE_LOCK(APCSG(lock));
 
@@ -234,6 +246,7 @@
 
     DESTROY_LOCK(APCSG(lock));
 }
+#endif
 
 #endif
 

Eigenschaftsänderungen: apc_string.c
___________________________________________________________________
Added: set
## -0,0 +1 ##
+svn:Id
\ No newline at end of property
Added: svn:keywords
## -0,0 +1 ##
+Id
\ No newline at end of property
Index: apc_bin.c
===================================================================
--- apc_bin.c	(Revision 311029)
+++ apc_bin.c	(Revision 324625)
@@ -28,6 +28,7 @@
 #include "apc_globals.h"
 #include "apc_bin.h"
 #include "apc_zend.h"
+#include "apc_php.h"
 #include "apc_sma.h"
 #include "apc_pool.h"
 #include "ext/standard/md5.h"
@@ -294,8 +295,6 @@
     apc_swizzle_hashtable(bd, ll, &ce->function_table, (apc_swizzle_cb_t)apc_swizzle_function, 0 TSRMLS_CC);
 #ifdef ZEND_ENGINE_2_4
     if (ce->default_properties_table) {
-        int i;
-
         for (i = 0; i < ce->default_properties_count; i++) {
             if (ce->default_properties_table[i]) {
                 apc_swizzle_ptr(bd, ll, &ce->default_properties_table[i]);
@@ -313,8 +312,6 @@
 
 #ifdef ZEND_ENGINE_2_4
     if (ce->default_static_members_table) {
-        int i;
-
         for (i = 0; i < ce->default_static_members_count; i++) {
             if (ce->default_static_members_table[i]) {
                 apc_swizzle_ptr(bd, ll, &ce->default_static_members_table[i]);
@@ -446,7 +443,7 @@
         zend_hash_index_update(&APCG(copied_zvals), (ulong)zv, (void**)&zv, sizeof(zval*), NULL);
     }
 
-    switch(zv->type & ~IS_CONSTANT_INDEX) {
+    switch(zv->type & IS_CONSTANT_TYPE_MASK) {
         case IS_NULL:
         case IS_LONG:
         case IS_DOUBLE:
Index: apc_string.h
===================================================================
--- apc_string.h	(Revision 311029)
+++ apc_string.h	(Revision 324625)
@@ -25,17 +25,19 @@
 
  */
 
-/* $Id: $ */
+/* $Id$ */
 
 #ifndef APC_STRING
 #define APC_STRING
 
 #include "apc.h"
 
+#ifndef ZTS
 void apc_interned_strings_init(TSRMLS_D);
 void apc_interned_strings_shutdown(TSRMLS_D);
+#endif
 
-char *apc_new_interned_string(char *arKey, int nKeyLength TSRMLS_DC);
+const char *apc_new_interned_string(const char *arKey, int nKeyLength TSRMLS_DC);
 
 #endif
 

Eigenschaftsänderungen: apc_string.h
___________________________________________________________________
Added: svn:keywords
## -0,0 +1 ##
+Id
\ No newline at end of property
Index: apc.c
===================================================================
--- apc.c	(Revision 311029)
+++ apc.c	(Revision 324625)
@@ -62,12 +62,13 @@
 
 void* apc_erealloc(void* p, size_t n TSRMLS_DC)
 {
-    p = realloc(p, n);
-    if (p == NULL) {
+    void *new;
+    new = realloc(p, n);
+    if (new == NULL) {
         apc_error("apc_erealloc: realloc failed to allocate %u bytes:" TSRMLS_CC, n);
         return NULL;
     }
-    return p;
+    return new;
 }
 
 void apc_efree(void* p TSRMLS_DC)
Index: php_apc.c
===================================================================
--- php_apc.c	(Revision 311029)
+++ php_apc.c	(Revision 324625)
@@ -1260,7 +1260,7 @@
 
             orig_current_execute_data = EG(current_execute_data);
             zend_try {
-                if (apc_compile_cache_entry(keys[i], &file_handle, ZEND_INCLUDE, t, &op_arrays[i], &cache_entries[i] TSRMLS_CC) != SUCCESS) {
+                if (apc_compile_cache_entry(&keys[i], &file_handle, ZEND_INCLUDE, t, &op_arrays[i], &cache_entries[i] TSRMLS_CC) != SUCCESS) {
                     op_arrays[i] = NULL;
                     cache_entries[i] = NULL;
                     add_assoc_long(return_value, Z_STRVAL_PP(hentry), -2);  /* -2: input or cache insertion error */
Index: apc_php.h
===================================================================
--- apc_php.h	(Revision 311029)
+++ apc_php.h	(Revision 324625)
@@ -63,6 +63,10 @@
 #include "zend_vm.h"
 #endif
 
+#ifndef IS_CONSTANT_TYPE_MASK
+#define IS_CONSTANT_TYPE_MASK (~IS_CONSTANT_INDEX)
+#endif
+
 #include "rfc1867.h"
 
 #endif
Index: tests/apc_011.phpt
===================================================================
--- tests/apc_011.phpt	(Revision 0)
+++ tests/apc_011.phpt	(Revision 324625)
@@ -0,0 +1,26 @@
+--TEST--
+APC: apc_fetch resets array pointers
+--SKIPIF--
+<?php require_once(dirname(__FILE__) . '/skipif.inc'); ?>
+--INI--
+apc.enabled=1
+apc.enable_cli=1
+apc.file_update_protection=0
+--FILE--
+<?php
+$items = array('bar', 'baz');
+
+apc_store('test', $items);
+
+$back = apc_fetch('test');
+
+var_dump(current($back));
+var_dump(current($back));
+
+?>
+===DONE===
+<?php exit(0); ?>
+--EXPECTF--
+string(3) "bar"
+string(3) "bar"
+===DONE===
Index: tests/apc_012.phpt
===================================================================
--- tests/apc_012.phpt	(Revision 0)
+++ tests/apc_012.phpt	(Revision 324625)
@@ -0,0 +1,98 @@
+--TEST--
+APC: apc_delete_file test with file_md5=on (see #61352)
+--SKIPIF--
+<?php require_once(dirname(__FILE__) . '/skipif.inc'); ?>
+--INI--
+apc.enabled=1
+apc.enable_cli=1
+apc.file_update_protection=0
+apc.stat=On
+apc.file_md5=1
+report_memleaks=0
+--FILE--
+<?php
+
+$files = array( 'apc_012.php',
+                'apc_009-1.php',
+                'apc_009-2.php',
+                'nofile.php',
+              );
+
+file_put_contents(dirname(__FILE__).'/apc_009-1.php', '<?php echo "test file";');
+file_put_contents(dirname(__FILE__).'/apc_009-2.php', '<?php syntaxerrorhere!');
+
+apc_compile_file($files[0]);
+check_file($files[0]);
+apc_delete_file($files[0]);
+check_file($files[0]);
+
+apc_compile_file($files[0]);
+apc_delete_file(array($files[0]));
+check_file($files[0]);
+
+apc_compile_file($files[0]);
+$it = new APCIterator('file');
+apc_delete_file($it);
+check_file($files[0]);
+
+var_dump(apc_compile_file(array($files[0], $files[1])));
+check_file(array($files[0], $files[1]));
+
+var_dump(apc_compile_file($files));
+check_file($files);
+
+function check_file($files) {
+
+  if (!is_array($files)) {
+    $files = array($files);
+  }
+
+  $info = apc_cache_info('file');
+
+  foreach ($files as $file) {
+    $match = 0;
+    foreach($info['cache_list'] as $cached_file) {
+      if (stristr($cached_file['filename'], $file)) $match = 1;
+    }
+    if ($match) {
+      echo "$file Found File\n";
+    } else {
+      echo "$file Not Found\n";
+    }
+  }
+}
+
+?>
+===DONE===
+<?php exit(0); ?>
+--CLEAN--
+<?php
+unlink('apc_009-1.php');
+unlink('apc_009-2.php');
+?>
+--EXPECTF--
+apc_012.php Found File
+apc_012.php Not Found
+apc_012.php Not Found
+apc_012.php Not Found
+array(0) {
+}
+apc_012.php Found File
+apc_009-1.php Found File
+
+Parse error: syntax error, unexpected '!' in %s/apc_009-2.php on line 1
+
+Warning: apc_compile_file(): Error compiling apc_009-2.php in apc_compile_file. in %s/apc_012.php on line 29
+
+Warning: apc_compile_file(): Error compiling nofile.php in apc_compile_file. in %s/apc_012.php on line 29
+array(2) {
+  ["apc_009-2.php"]=>
+  int(-1)
+  ["nofile.php"]=>
+  int(-1)
+}
+apc_012.php Found File
+apc_009-1.php Found File
+apc_009-2.php Not Found
+nofile.php Not Found
+===DONE===
Index: apc_cache.c
===================================================================
--- apc_cache.c	(Revision 311029)
+++ apc_cache.c	(Revision 324625)
@@ -314,7 +314,7 @@
 void apc_cache_destroy(apc_cache_t* cache TSRMLS_DC)
 {
     DESTROY_LOCK(cache->header->lock);
-#ifdef NONBLOCKING_LOCK_AVAILABLE
+#if NONBLOCKING_LOCK_AVAILABLE
     DESTROY_LOCK(cache->header->wrlock);
 #endif
     apc_efree(cache TSRMLS_CC);
@@ -1290,7 +1290,9 @@
                 add_next_index_zval(list, link);
                 j++;
             }
-            add_next_index_long(slots, j);
+            if(j != 0) {
+                add_index_long(slots, (ulong)i, j);
+            }
         }
 
         /* For each slot pending deletion */
Index: apc_cache.h
===================================================================
--- apc_cache.h	(Revision 311029)
+++ apc_cache.h	(Revision 324625)
@@ -287,7 +287,7 @@
                                                     TSRMLS_DC);
 
 
-zend_bool apc_compile_cache_entry(apc_cache_key_t key, zend_file_handle* h, int type, time_t t, zend_op_array** op_array_pp, apc_cache_entry_t** cache_entry_pp TSRMLS_DC);
+zend_bool apc_compile_cache_entry(apc_cache_key_t *key, zend_file_handle* h, int type, time_t t, zend_op_array** op_array_pp, apc_cache_entry_t** cache_entry_pp TSRMLS_DC);
 
 /*
  * apc_cache_make_user_entry creates an apc_cache_entry_t object given an info string
Index: config.m4
===================================================================
--- config.m4	(Revision 311029)
+++ config.m4	(Revision 324625)
@@ -177,7 +177,7 @@
 	LIBS="$orig_LIBS"
 fi
 
-	AC_CACHE_CHECK([whether the target compiler supports builtin atomics], PHP_APC_GCC_ATOMICS, [
+	AC_CACHE_CHECK([whether the target compiler supports builtin atomics], PHP_cv_APC_GCC_ATOMICS, [
 
 			AC_TRY_LINK([],[
 					int foo = 0;
@@ -185,11 +185,11 @@
 					__sync_bool_compare_and_swap(&foo, 0, 1);
 					return __sync_fetch_and_add(&foo, 1);
 				],
-				[PHP_APC_GCC_ATOMICS=yes],
-				[PHP_APC_GCC_ATOMICS=no])
+				[PHP_cv_APC_GCC_ATOMICS=yes],
+				[PHP_cv_APC_GCC_ATOMICS=no])
 		])
 
-	if test "x${PHP_APC_GCC_ATOMICS}" != "xno"; then
+	if test "x${PHP_cv_APC_GCC_ATOMICS}" != "xno"; then
 			AC_DEFINE(HAVE_ATOMIC_OPERATIONS, 1,
 				[Define this if your target compiler supports builtin atomics])
 		else
Index: apc_lock.h
===================================================================
--- apc_lock.h	(Revision 311029)
+++ apc_lock.h	(Revision 324625)
@@ -154,7 +154,7 @@
 # define apc_lck_nb_lock(a)    apc_fcntl_nonblocking_lock(a TSRMLS_CC)
 # define apc_lck_rdlock(a)     apc_fcntl_rdlock(a TSRMLS_CC)
 # define apc_lck_unlock(a)     apc_fcntl_unlock(a TSRMLS_CC)
-# define apc_lck_rdunlock(a)   apc_fcntl_unlock(&a TSRMLS_CC)
+# define apc_lck_rdunlock(a)   apc_fcntl_unlock(a TSRMLS_CC)
 #endif
 
 #endif
